### PR TITLE
Deploy shared dev key vault for mock MSI certificates

### PR DESCRIPTION
**Pull request by kimorris27**
### Which issue this PR addresses:

https://issues.redhat.com/browse/ARO-9005

### What this PR does / why we need it:

This PR:

1. Adds a new key vault to the shared dev infra for persisting mock cluster MSI certificates
2. Adds a role assignment that grants the mock RP identity permission to manage secrets in the key vault
3. Renames the preexisting shared dev key vault for disk encryption to clarify its purpose

It also includes the contents of #3687, since I found that issue while working on these changes.

### Test plan for issue:

I deployed to our shared dev infra in `westeurope` using `deploy_env_dev` and everything looked good.

### Is there any documentation that needs to be updated for this PR?

N/A

### How do you know this will function as expected in production? 

N/A - this change only affects local dev mode

### CODE REVIEW

File: pkg/deploy/generator/resources_dev.go [hunk 1]
```diff
@@ -445,3 +445,25 @@ func (g *generator) devDiskEncryptionKeyVaultAccessPolicy() *arm.Resource {
 		DependsOn:  []string{fmt.Sprintf("[resourceId('Microsoft.Compute/diskEncryptionSets', %s)]", sharedDiskEncryptionSetName)},
 	}
 }
+
+const (
+	SharedMSIKeyVaultName       = "concat(take(resourceGroup().name,10), '" + SharedMSIKeyVaultNameSuffix + "')"
```
Comment (rajdeepc2792) [inline pkg/deploy/generator/resources_dev.go:450]: Can we use all the available length fot the keyvault name? I believe 10 was used in case of shared kv just to keep it to 23.

Comment (rajdeepc2792) [reply pkg/deploy/generator/resources_dev.go:450]: Taking this comment back, it should follow the same assumption as the other KV otherwise in the RP backend it may need an additional handling:- https://github.com/Azure/ARO-RP/blob/master/pkg/util/cluster/cluster.go#L218C2-L224C3

**Review by rajdeepc2792 — COMMENTED** [review]
LGTM, just wanted to ask about how the Full RP is going to the mocking for the same?

**Review by rajdeepc2792 — DISMISSED** [review]

**Comment by kimorris27** [comment]
> LGTM, just wanted to ask about how the Full RP is going to the mocking for the same?

I think everything should be the same, since what we call a "full service dev RP" is essentially a shared dev RP that's only being used by one person.

**Comment by kimorris27** [comment]
I talked with Rajdeep a bit in Slack. We concluded that this does need some additional work to function properly in the context of a full service dev RP, since the role assignment on the key vault needs to be granted to the MI associated with the RP VMSS rather than the mock identity we use with our shared dev RP infrastructure.

I'm going to follow the same pattern Rajdeep did with the OIC storage account where the OIC stuff is in a separate ARM template so that the RP SP ID can be passed in as a parameter. I'll ask for review again once I've made these changes and tested by deploying a full service RP up to the point where these things get created.

**Comment by SudoBrendan** [comment]
Looks like 3687 has merged, can we rebase to minimize the diff?

**Review by SudoBrendan — APPROVED** [review]

**Review by cadenmarchese — APPROVED** [review]

File: pkg/deploy/generator/templates_msi_kv.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package generator
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+)
+
+func (g *generator) devMSIKeyvaultTemplate() *arm.Template {
```
[the remark this answers is not in the document]
Comment (kimorris27) [reply pkg/deploy/generator/templates_msi_kv.go:10]: Good point. I think the two should always go together. I'll combine the two before we merge, and I'll give them a name indicating that the template contains infra needed for MIWI clusters. I'll skip another full service RP test though for time's sake.

[the remark this answers is not in the document]
Comment (kimorris27) [reply pkg/deploy/generator/templates_msi_kv.go]: Consolidated the ARM templates and validated the new ARM template by testing out `deploy_miwi_infra_dev`.

**Review by cadenmarchese — APPROVED** [review]
nice, thank you!

**Review by mociarain — COMMENTED** [review]
LGTM but I ahve similar question about naming I raised here:https://github.com/Azure/ARO-RP/pull/3713#pullrequestreview-2193133644

tl;dr Should we be using `MI` over `MSI` to refer to this stuff?

**Review by rajdeepc2792 — APPROVED** [review]
LGTM

**Comment by kimorris27** [comment]
>Should we be using MI over MSI to refer to this stuff?

@mociarain - Technically speaking, the answer is probably yes. But I personally like using "cluster MSI" when referring to _that particular MI that the RP leverages to act on behalf of the customer_ because there are so many MIs everywhere that it's useful to have a term for this particular one. "Managed service identity" happens to work nicely here IMO.
